### PR TITLE
avoid using LocalFree on FormatMessageW buffer

### PR DESCRIPTION
**Pull request by ryanlai2**
LocalFree doesn't work downlevel on windows 8.1. 

Onecoreuap.lib should be used so that specific apisets don't need to be required to run on downlevel OS's like 8.1 and can fall back to using kernel32.dll using reverseforwarders.

The builds need to be changed in a separate change, from using Onecoreuap_apiset.lib to using onecoreuap.lib. With this change, a check must be added in the CI pipelines by using the APIValidator tool to check for regressions

### CODE REVIEW

File: onnxruntime/core/platform/windows/env.cc [hunk 1]
```diff
@@ -544,22 +544,21 @@ class WindowsEnv : public Env {
 #endif
     if (!*handle) {
       const auto error_code = GetLastError();
-      LPVOID lpMsgBuf;
+      static const DWORD bufferLength = 64 * 1024;
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env.cc:547]: why 64*1024?

Comment (ryanlai2) [reply onnxruntime/core/platform/windows/env.cc:547]: I was referencing this PR : https://github.com/microsoft/onnxruntime/pull/3772/files

What do you suggest?

**Comment by snnn** [comment]
I think there are two conflict requirements:
1. Have a build that can support all Windows 10/11 variants
2. Have a build that can support all downlevel Windows desktop versions starting from Windows 7.

I believe Onecoreuap.lib and Onecoreuap_apiset.lib play different roles here. Am I right?

**Comment by ryanlai2** [comment]
> I think there are two conflict requirements:

> 1. Have a build that can support all Windows 10/11 variants. 

Yes, correct.

> 2. Have a build that can support all downlevel Windows desktop versions starting from Windows 7.
The lowest one is Windows 8.1
> I believe Onecoreuap.lib and Onecoreuap_apiset.lib play different roles here. Am I right?

We are trying to produce binaries that work both downlevel on desktop down to Win 8.1, and across WCOS devices.

Onecoreuap.lib and onecoreuap_apiset.lib have the same API coverage. However Onecoreuap.lib will make it so that apiset dlls are not explicitly required and will fallback to legacy .dll's using Reverse Forwarders. https://docs.microsoft.com/en-us/windows/win32/apiindex/api-set-loader-operation#reverse-forwarding

Right now with onecoreuap_apiset.lib,  direct forwarding is used and the api set dll is attempted to be loaded directly and the application crashes if that apiset dll doesn't exist such as on 8.1. 

In this case, LocalFree is available on Kernel32.dll but it chooses to use api-ms-win-core-heap-l2-1-0.dll

MicrosoftTeams-image

Addressing your concern about validation, API Validator can be added to the pipeline so that it'll fail / emit warnings if API's are used in Onnxruntime.dll and Winml dll are breaking backwards compatibility : https://docs.microsoft.com/en-us/windows-hardware/drivers/develop/validating-windows-drivers#apivalidator

Transitioning to using onecoreuap.lib + adding APIValidator into the pipeline may take some time. I think that work definitely needs to be done to make developer lives easier since we'll have pipeline level validation. However, since 1.11 release deadline is coming up, I would like to add this PR's change to unblock Windows 8.1 and in a separate PR update to use onecoreuap.lib + API validator.

What do you think?

**Review by snnn — APPROVED** [review]

**Comment by snnn** [comment]
[ryanlai2](https://github.com/ryanlai2), [smk2007](https://github.com/smk2007) and I had a discussion: for this release, we will keep it as what it was for the last release, which is we should avoid using LocalFree.  And @ryanlai2 will use APIValidator to verify it. 

Going forward, we need try to make it simpler.  Now it works because Tiago added some hacks. I suggested a single desktop binaries should be enough for more use cases. For example, Windows python.exe is such a things. Python.exe links to kernel32.dll but it works fine on onecore and onecoreuap editions. I suggest onnxruntime does the same, and be aligned with what Visual Studio provides. onnxruntime  should keep the wcos build option in build.py in case someone still wants a pure onecore build.

**Review by snnn — APPROVED** [review]

**Comment by hawkhai** [comment]
You can implement it yourself to support Windows 7. The issue is resolved here:
可以自己实现，以支持 win7，这里解决了：<https://github.com/yycmagic/onnxruntime-for-win7>

File: onnxruntime/core/platform/windows/env.cc [hunk 2]
```diff
@@ -544,22 +544,21 @@ class WindowsEnv : public Env {
...
       FormatMessageW(
-          FORMAT_MESSAGE_ALLOCATE_BUFFER |
               FORMAT_MESSAGE_FROM_SYSTEM |
               FORMAT_MESSAGE_IGNORE_INSERTS,
           NULL,
           error_code,
           MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
-          (LPWSTR)&lpMsgBuf,
+          (LPWSTR)s.data(),
           0, NULL);
```
Comment (snnn) [inline onnxruntime/core/platform/windows/env.cc:556]: Here you need to pass in the size of the buffer

**Pull request merged**